sourced/ml/transformers/content2ids.py

```diff
@@ -0,0 +1,146 @@
+        self.documents_column = documents_column
+        self.linguist2pygments = {}
+        self.split = args.split
+        self.idfreq = args.idfreq
+        self.output = args.output
+
+    def __call__(self, rows):
+        list_RDDs = []
+        self.build_mapping()
+        processed = rows.flatMap(self._process_row).persist()
+
+        if self.idfreq:
+            num_repos_processed = processed \
+                .map(lambda x: (x[0], x[1][0])) \
+                .distinct()
+            num_repos_reduced = self.reduce_rows(num_repos_processed)
+            list_RDDs.append(num_repos_reduced)
+
+            num_files_processed = processed \
+                .map(lambda x: (x[0], x[1][1])) \
+                .distinct()
+            num_files_reduced = self.reduce_rows(num_files_processed)
+            list_RDDs.append(num_files_reduced)
+
+            num_occ_reduced = self.reduce_rows(processed)
+            list_RDDs.append(num_occ_reduced)
+
+            return processed \
+                .map(lambda x: x[0]) \
+                .context.union(list_RDDs) \
+                .groupByKey() \
+                .mapValues(list) \
+                .map(lambda x: Row(
+                    token=x[0],
+                    token_split=" ".join(TokenParser(min_split_length=1).split(x[0])),
+                    num_repos=x[1][0],
+                    num_files=x[1][1],
+                    num_occ=x[1][2]))
+        else:
+            return processed \
+                .map(lambda x: x[0]) \
+                .distinct() \
+                .map(lambda x: Row(
+                    token=x,
+                    token_split=" ".join(TokenParser(min_split_length=1).split(x))))
+
+    def reduce_rows(self, rows):
+        return rows \
+            .map(lambda x: (x[0], 1)) \
+            .reduceByKey(operator.add)
+
+    def _process_row(self, row):
+        self.names = []
+        repo_id = getattr(row, self.documents_column[0])
+        file_id = getattr(row, self.documents_column[1])
+        path = os.path.join(repo_id, file_id)
+        code = row.content
+        try:
+            lexer = get_lexer_by_name(self.linguist2pygments[row.lang][0])
+            pygments.highlight(code, lexer, self.FormatterProxy(callback=self.process_tokens))
+        except ClassNotFound:
+            lexer = get_lexer_by_name(self.linguist2pygments[row.lang][1])
+            pygments.highlight(code, lexer, self.FormatterProxy(callback=self.process_tokens))
+        except KeyError:
+            pass
+        for token in self.names:
+            yield token, (repo_id, path)
+
+    def process_tokens(self, tokens):
+        """
+        Filter tokens of type "Name" and which are splittable
+        according to :class: 'TokenParser' rules
+        """
+        for _type, token in tokens:
+            if _type[0] == "Name":
+                if self.split:
+                    if len(list(TokenParser(min_split_length=1).split(token))) > 1:
```

Comment:
`len` + `list` is a bad way to calc length of iterators. `sum(1 for _ in ...)` is much better because it does not materialize all the items